tensorflow_probability/python/experimental/stats/sample_stats.py

```diff
@@ -0,0 +1,246 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""Functions for computing statistics of samples."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+# Dependency imports
+
+import collections
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.internal import prefer_static as ps
+
+
+__all__ = [
+    'RunningCovariance',
+    'RunningCovarianceState',
+    'RunningVariance',
+]
+
+
+RunningCovarianceState = collections.namedtuple(
+    'RunningCovarianceState', 'num_samples, mean, comoment')
+
+
+def _update_running_covariance(
+    state, new_sample, event_ndims, dtype, axis
+):
+  new_sample = tf.cast(new_sample, dtype=dtype)
+  if axis is not None:
+    chunk_n = tf.cast(ps.shape(new_sample)[axis], dtype=dtype)
+    chunk_mean = tf.math.reduce_mean(new_sample, axis=axis)
+    chunk_delta_mean = new_sample - tf.expand_dims(chunk_mean, axis=axis)
+    chunk_comoment = tf.reduce_sum(
+        _batch_outer_product(chunk_delta_mean, event_ndims),
+        axis=axis
+    )
+  else:
+    chunk_n = tf.ones((), dtype=dtype)
+    chunk_mean = new_sample
+    chunk_comoment = tf.zeros(ps.shape(state.comoment), dtype=dtype)
+
+  new_n = state.num_samples + chunk_n
+  delta_mean = chunk_mean - state.mean
+  new_mean = state.mean + chunk_n * delta_mean / new_n
+  all_pairwise_deltas = _batch_outer_product(
+      delta_mean, event_ndims)
+  adj_factor = state.num_samples * chunk_n / (state.num_samples + chunk_n)
+  new_comoment = (state.comoment + chunk_comoment
+                  + adj_factor * all_pairwise_deltas)
+
+  return type(state)(new_n, new_mean, new_comoment)
+
+def _batch_outer_product(target, event_ndims):
+  """Calculates the batch outer product along `target`'s event dimensions.
+
+  More precisely, A `tf.einsum` operation is used to calculate desired
+  pairwise products as follows:
+
+  For `event_ndims=0`, the return value is:
+    `tf.einsum("...,...->...", target, target)`
+  For `event_ndims=1`:
+    `tf.einsum("...a,...b->...ab", target, target)`
+  For `event_ndims=2`:
+    `tf.einsum("...ab,...cd->...abcd", target, target)`
+  ...
+
+  Args:
+    target: Target `Tensor` for the `tf.einsum` computation.
+    event_ndims: Both the number of dimensions that specify the event
+      shape and the desired number of dimensions for cross product
+      terms.
+
+  Returns:
+    outer_product: A `Tensor` with shape B + E + E for all pairwise
+      products of `target` in the event dimensions.
+  """
+  assign_indices = ''.join(list(
+      map(chr, range(ord('a'), ord('a')+event_ndims*2))))
```

Comment:
There should be spaces around binary operators (+ and * here)

Reply:
Good catch!